test/test_distributed.py

```diff
@@ -700,13 +734,79 @@ def test_all_gather_multigpu(self):
         self._test_all_gather_multigpu_helper(group, group_id, rank,
                                               rankToGPUMapping)
 
+    # END TO END TEST FOR DISTRIBUTEDDATAPARALLEL
+    def _test_DDP_helper(self, model, input_var, target, loss, optimizer):
+        model.train()
+        output = model(input_var)
+        l = loss(output, target)
+        l.backward()
+        optimizer.step()
+        return model
+
+    @unittest.skipIf(BACKEND != 'nccl' and BACKEND != 'gloo',
+                     "Only Nccl & Gloo backend support DistributedDataParallel")
+    @skip_if_no_cuda_distributed
+    @skip_if_no_multigpu
+    def test_DistributedDataParallel(self):
+        # Run a simple end to end DDP model, use result of single node model
+        # as baseline
+        group, group_id, rank = self._init_global_test()
+        rankToGPUMapping = self._init_multigpu_helper()
+
+        class Net(nn.Module):
+            def __init__(self):
+                super(Net, self).__init__()
+                self.features = nn.Linear(2, 4, bias=False)
+
+            def forward(self, x):
+                x = self.features(x)
+                return x
+
+        # cpu training setup
+        model = Net()
+        model_cpu = copy.deepcopy(model)
+        optimizer_cpu = torch.optim.SGD(model_cpu.parameters(), 0.1)
+
+        # DDP training setup
+        model_DDP = copy.deepcopy(model)
+        gpu_subset = list(rankToGPUMapping[rank])
+        model_DDP.cuda(gpu_subset[0])
+        model_DDP = nn.parallel.DistributedDataParallel(model_DDP,
+                                                        device_ids=gpu_subset)
+        optimizer_DDP = torch.optim.SGD(model_DDP.parameters(), 0.1)
```

Comment:
It would be better to skip the optimizer and compare `.grad`s directly (it only adds complexity unnecessarily)